core/src/main/java/org/opensearch/sql/expression/function/SpanFunctionImpl.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.calcite.schema.impl.ScalarFunctionImpl;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.util.BuiltInMethod;
+import org.opensearch.sql.calcite.type.ExprSqlType;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.data.type.ExprCoreType;
+import org.opensearch.sql.planner.physical.collector.Rounding;
+import org.opensearch.sql.planner.physical.collector.Rounding.DateRounding;
+import org.opensearch.sql.planner.physical.collector.Rounding.TimeRounding;
+import org.opensearch.sql.planner.physical.collector.Rounding.TimestampRounding;
+
+public class SpanFunctionImpl extends ImplementorUDF {
+  protected SpanFunctionImpl() {
+    super(new SpanImplementor());
+  }
+
+  @Override
+  public SqlReturnTypeInference getReturnTypeInference() {
+    return ReturnTypes.ARG0;
+  }
+
+  public static class SpanImplementor implements NotNullImplementor {
+    @Override
+    public Expression implement(
+        RexToLixTranslator translator, RexCall call, List<Expression> translatedOperands) {
+      assert call.getOperands().size() == 3;
+      assert translatedOperands.size() == 3;
+      Expression field = translatedOperands.get(0);
+      Expression interval = translatedOperands.get(1);
+
+      RelDataType fieldType = call.getOperands().get(0).getType();
+      RelDataType unitType = call.getOperands().get(2).getType();
+
+      if (SqlTypeUtil.isNull(unitType)) {
+        return switch (call.getType().getSqlTypeName()) {
+          case BIGINT, INTEGER, SMALLINT, TINYINT -> Expressions.multiply(
+              Expressions.divide(field, interval), interval);
+          default -> Expressions.multiply(
+              Expressions.call(BuiltInMethod.FLOOR.method, Expressions.divide(field, interval)),
+              interval);
+        };
+      } else if (fieldType instanceof ExprSqlType exprSqlType) {
+        // TODO: pass in constant arguments when constructing
+        String methodName =
+            switch (exprSqlType.getUdt()) {
+              case EXPR_DATE -> "evalDate";
+              case EXPR_TIME -> "evalTime";
+              case EXPR_TIMESTAMP -> "evalTimestamp";
+              default -> throw new IllegalArgumentException(
+                  String.format("Unsupported expr udt: %s", exprSqlType.getUdt()));
+            };
+        ScalarFunctionImpl function =
+            (ScalarFunctionImpl)
+                ScalarFunctionImpl.create(
+                    Types.lookupMethod(
+                        SpanFunctionImpl.class, methodName, String.class, int.class, String.class));
+        return function.getImplementor().implement(translator, call, RexImpTable.NullAs.NULL);
+      }
+      throw new IllegalArgumentException(String.format("Unsupported RelDataType: %s", fieldType));
```

Comment:
RelDataType is internal implementation, we should hide internal error message from user.